dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Tearing.java

```diff
@@ -0,0 +1,308 @@
+        final Expression filter = program.getFilterSpecification();
+        if (specification == null && filter == null) {
+            return;
+        }
+        final var substitution = new FinalValueTearSubstitution();
+        for (Init init : program.getThreadEvents(Init.class)) {
+            substitution.typesByObject.computeIfAbsent(init.getBase(), k -> new HashMap<>())
+                    .put(init.getOffset(), init.getAccessType());
+        }
+        final Expression updatedSpecification = specification == null ? null : specification.accept(substitution);
+        final Expression updatedFilter = filter == null ? null : filter.accept(substitution);
+        program.setSpecification(program.getSpecificationType(), updatedSpecification);
+        program.setFilterSpecification(updatedFilter);
+    }
+
+    private List<Event> createTransaction(Load load, List<Integer> offsets) {
+        final int bytes = checkBytes(load, offsets);
+        final List<Event> replacement = new ArrayList<>();
+        final IntegerType addressType = checkIntegerType(load.getAddress().getType(),
+                "Non-integer address in '%s'", load);
+        checkIntegerType(load.getAccessType(), "Non-integer mixed-size access in '%s'", load);
+        final Function function = load.getFunction();
+        final Register addressRegister = toRegister(load.getAddress(), function, replacement);
+        final List<Register> smallerRegisters = new ArrayList<>();
+        for (int i = -1; i < offsets.size(); i++) {
+            int start = i < 0 ? 0 : offsets.get(i);
+            int end = i + 1 < offsets.size() ? offsets.get(i + 1) : bytes;
+            assert start < end;
+            smallerRegisters.add(function.newRegister(types.getIntegerType(8 * (end - start))));
+        }
+        assert bytes == smallerRegisters.stream().mapToInt(t -> types.getMemorySizeInBytes(t.getType())).sum();
+        final InstructionBoundary begin = load.hasTag(Tag.NO_INSTRUCTION) ? null : EventFactory.newInstructionBegin();
+        if (begin != null) {
+            replacement.add(begin);
+        }
+        for (int i = -1; i < offsets.size(); i++) {
+            final int start = i < 0 ? 0 : offsets.get(i);
+            final Expression offset = expressions.makeValue(start, addressType);
+            final Expression address = expressions.makeAdd(addressRegister, offset);
+            final Load byteLoad = load.getCopy();
+            final Register result = smallerRegisters.get(i + 1);
+            byteLoad.setResultRegister(result);
+            byteLoad.setAccessType(result.getType());
+            byteLoad.setAddress(address);
+            replacement.add(byteLoad);
+        }
+        if (begin != null) {
+            replacement.add(EventFactory.newInstructionEnd(begin));
+        }
+        final Expression combination = expressions.makeIntConcat(smallerRegisters);
+        replacement.add(EventFactory.newLocal(load.getResultRegister(), combination));
```

Comment:
I think this instruction should copy the source location information of the original load (the instruction boundary events should probably too). The same applies to the store case.